service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalog.java

```diff
@@ -1328,6 +1347,38 @@ public void doCommit(TableMetadata base, TableMetadata metadata) {
       String newLocation = writeNewMetadataIfRequired(base == null, metadata);
       String oldLocation = base == null ? null : base.metadataFileLocation();
 
+      // TODO: we should not need to do this hack, but there's no other way to modify
+      // currentMetadata / currentMetadataLocation
+      if (updateMetadataOnCommit) {
+        try {
+          Field tableMetadataField = TableMetadata.class.getDeclaredField("metadataFileLocation");
+          tableMetadataField.setAccessible(true);
+          tableMetadataField.set(metadata, newLocation);
+
+          Field unsafeField = Unsafe.class.getDeclaredField("theUnsafe");
+          unsafeField.setAccessible(true);
+          Unsafe unsafe = (Unsafe) unsafeField.get(null);
+          Field changesField = TableMetadata.class.getDeclaredField("changes");
+          changesField.setAccessible(true);
+          long offset = unsafe.objectFieldOffset(changesField);
+          unsafe.putObject(metadata, offset, new ArrayList<MetadataUpdate>());
+
+          Field currentMetadataLocationField =
+              BaseMetastoreTableOperations.class.getDeclaredField("currentMetadataLocation");
+          currentMetadataLocationField.setAccessible(true);
+          currentMetadataLocationField.set(this, newLocation);
```

Comment:
Forcing the new location may be ok on the assumption that the commit succeeds, but can we be sure at this point that the commit will succeed?

Reply:
Well, TBH, that statement makes me worry more about the correctness of this change :sweat_smile: Could you give more details?

Reply:
Well maybe not worry but at least it's good we are being cautious. The bug and the fix here are both subtle.

What I mean is, once the metadata is actually written to the metadata location, it should be correct to attach the location to it. If the commit ends up failing, it seems like we (in the tests, at least) don't make any assumptions that the metadata is still useful after the failed commit. So the location of this in the method should not matter. If anything, I would think that the TableOperations pointing at a "failed" commit could be troublesome, but previously it was just pointing at null.